File: app/src/main/java/com/dumuzeyn/mp3player/MainRenderer.java
package com.dumuzeyn.mp3player;

import android.view.View;
import android.view.ViewTreeObserver;
import android.widget.ScrollView;
import java.util.HashMap;
import java.util.Map;

final class MainRenderer {
    private final MainActivityCore host;
    private final MenuRenderer homeRenderer;
    private final SongsMenuRenderer songsRenderer;
    private final FavoritesMenuRenderer favoritesRenderer;
    private final PlaylistsMenuRenderer playlistsRenderer;
    private final MenuRenderer soundRenderer;
    private final MenuRenderer genresRenderer;
    private final MenuRenderer artistsRenderer;
    private final MenuRenderer albumsRenderer;
    private final MenuRenderer settingsRenderer;
    private final MenuRenderer foldersRenderer;
    private final Map<String, Integer> scrollPositions = new HashMap<>();
    private String renderedMenuKey;

    MainRenderer(MainActivityCore host) {
        this.host = host;
        this.homeRenderer = new HomeMenuRenderer(host);
        this.songsRenderer = new SongsMenuRenderer(host);
        this.favoritesRenderer = new FavoritesMenuRenderer(host);
        this.playlistsRenderer = new PlaylistsMenuRenderer(host);
        this.soundRenderer = new SoundMenuRenderer(host);
        this.genresRenderer = new GenresMenuRenderer(host);
        this.artistsRenderer = new ArtistsMenuRenderer(host);
        this.albumsRenderer = new AlbumsMenuRenderer(host);
        this.settingsRenderer = new SettingsMenuRenderer(host);
        this.foldersRenderer = new FoldersMenuRenderer(host);
    }

    void render() {
        rememberCurrentScrollPosition();
        host.refreshTabs();
        host.navigationState.songRenderGeneration++;
        if (host.navigationState.tabIndex == LibraryTabs.SONGS
                && !host.navigationState.renderingTabPreview
                && host.songsView != null) {
            host.songRows.clear();
            host.sourcePlayButton = null;
            if (host.contentScroll != null) {
                host.contentScroll.setVisibility(View.GONE);
            }
            songsRenderer.render();
            renderedMenuKey = menuKey(
                    host.navigationState.tabIndex, host.navigationState.search);
            host.playerUiController.updateMini();
            return;
        }
        if (host.songsView != null) {
            host.songsView.hide();
        }
        if (host.contentScroll != null) {
            host.contentScroll.setVisibility(View.VISIBLE);
        }
        host.list.removeAllViews();
        host.songRows.clear();
        host.sourcePlayButton = null;
        host.renderSectionHeader();
        MenuRenderer renderer = rendererForTab();
        int scrollY = scrollPositionFor(
                host.navigationState.tabIndex, host.navigationState.search);
        if (host.navigationState.tabIndex == LibraryTabs.SONGS
                || host.navigationState.tabIndex == LibraryTabs.FAVORITES) {
            host.songsRenderer.prepareNextRenderForScroll(scrollY);
        }
        renderer.render();
        if (renderer.needsMiniSpacer()) {
            host.addMiniSpacerIfNeeded();
        }
        restoreCurrentScrollPosition();
        host.playerUiController.updateMini();
    }

    void captureScrollBeforeUiRebuild() {
        rememberCurrentScrollPosition();
        renderedMenuKey = null;
    }

    void captureCurrentScrollPosition() {
        rememberCurrentScrollPosition();
    }

    private void rememberCurrentScrollPosition() {
        if (renderedMenuKey == null || host.contentScroll == null) {
            return;
        }
        if (renderedMenuKey.startsWith(LibraryTabs.SONGS + "\n")) {
            return;
        }
        scrollPositions.put(renderedMenuKey, Math.max(0, host.contentScroll.getScrollY()));
    }

    private void restoreCurrentScrollPosition() {
        renderedMenuKey = menuKey(host.navigationState.tabIndex, host.navigationState.search);
        final String targetKey = renderedMenuKey;
        if (host.contentScroll == null) {
            return;
        }
        int scrollY = scrollPositions.containsKey(renderedMenuKey)
                ? scrollPositions.get(renderedMenuKey) : 0;
        final ScrollView targetScroll = host.contentScroll;
        if (scrollY <= 0) {
            targetScroll.scrollTo(0, 0);
            targetScroll.setVisibility(View.VISIBLE);
            return;
        }
        targetScroll.setVisibility(View.INVISIBLE);
        targetScroll.getViewTreeObserver().addOnPreDrawListener(
                new ViewTreeObserver.OnPreDrawListener() {
                    @Override
                    public boolean onPreDraw() {
                        ViewTreeObserver observer = targetScroll.getViewTreeObserver();
                        if (observer.isAlive()) {
                            observer.removeOnPreDrawListener(this);
                        }
                        if (host.contentScroll != targetScroll
                                || !targetKey.equals(menuKey(host.navigationState.tabIndex, host.navigationState.search))) {
                            targetScroll.setVisibility(View.VISIBLE);
                            return true;
                        }
                        targetScroll.scrollTo(0, scrollY);
                        targetScroll.setVisibility(View.VISIBLE);
                        return false;
                    }
                });
    }

    private String menuKey(int tabIndex, String search) {
        return tabIndex + "\n" + (search == null ? "" : search);
    }

    PreviewState renderPreview(android.widget.LinearLayout target, int targetIndex, String targetSearch) {
        android.widget.LinearLayout previousList = host.list;
        ButtonState previousButton = new ButtonState(host.sourcePlayButton);
        SongsRenderer.BatchState previousBatchState = host.songsRenderer.captureBatchState();
        int previousTab = host.navigationState.tabIndex;
        int previousGeneration = host.navigationState.songRenderGeneration;
        String previousSearch = host.navigationState.search;
        boolean previousPreview = host.navigationState.renderingTabPreview;
        int scrollY = scrollPositionFor(targetIndex, targetSearch);
        PreviewState previewState;
        try {
            host.previewSongRows.clear();
            host.list = target;
            host.navigationState.tabIndex = targetIndex;
            host.navigationState.songRenderGeneration = previousGeneration + 1;
            host.navigationState.search = targetSearch == null ? "" : targetSearch;
            host.navigationState.renderingTabPreview = true;
            host.sourcePlayButton = null;
            target.removeAllViews();
            host.renderSectionHeader();
            MenuRenderer renderer = rendererForTab(targetIndex);
            if (targetIndex == LibraryTabs.SONGS
                    || targetIndex == LibraryTabs.FAVORITES) {
                host.songsRenderer.prepareNextRenderForScroll(scrollY);
            }
            renderer.render();
            if (renderer.needsMiniSpacer()) {
                host.addMiniSpacerIfNeeded();
            }
            previewState = new PreviewState(
                    scrollY,
                    host.navigationState.songRenderGeneration,
                    host.songsRenderer.captureBatchState(),
                    host.sourcePlayButton);
        } finally {
            host.list = previousList;
            host.navigationState.tabIndex = previousTab;
            host.navigationState.songRenderGeneration = previousGeneration;
            host.navigationState.search = previousSearch;
            host.navigationState.renderingTabPreview = previousPreview;
            host.sourcePlayButton = previousButton.button;
            host.songsRenderer.restoreBatchState(previousBatchState);
        }
        return previewState;
    }

    void adoptPreview(int targetIndex, String targetSearch, PreviewState state) {
        if (host.songsView != null) {
            host.songsView.hide();
        }
        renderedMenuKey = menuKey(targetIndex, targetSearch);
        host.navigationState.songRenderGeneration = state.generation;
        host.songsRenderer.restoreBatchState(state.batchState);
        host.songRows.replaceWith(host.previewSongRows);
        host.previewSongRows.clear();
        host.sourcePlayButton = state.sourcePlayButton;
        host.songRows.refresh(host.songRowStateResolver());
        host.artworkUi.promoteVisibleArtwork();
        host.playerUiController.updateMini();
    }

    void discardPreview() {
        host.previewSongRows.clear();
    }

    private int scrollPositionFor(int tabIndex, String search) {
        Integer position = scrollPositions.get(menuKey(tabIndex, search));
        return position == null ? 0 : Math.max(0, position);
    }

    private MenuRenderer rendererForTab() {
        return rendererForTab(host.navigationState.tabIndex);
    }

    private MenuRenderer rendererForTab(int tabIndex) {
        if (tabIndex == LibraryTabs.HOME) {
            return homeRenderer;
        }
        if (tabIndex == LibraryTabs.SONGS) {
            return songsRenderer;
        }
        if (tabIndex == LibraryTabs.FAVORITES) {
            return favoritesRenderer;
        }
        if (tabIndex == LibraryTabs.PLAYLISTS) {
            return playlistsRenderer;
        }
        if (tabIndex == LibraryTabs.SOUND) {
            return soundRenderer;
        }
        if (tabIndex == LibraryTabs.GENRES) {
            return genresRenderer;
        }
        if (tabIndex == LibraryTabs.ARTISTS) {
            return artistsRenderer;
        }
        if (tabIndex == LibraryTabs.ALBUMS) {
            return albumsRenderer;
        }
        if (tabIndex == LibraryTabs.FOLDERS) {
            return foldersRenderer;
        }
        if (tabIndex == LibraryTabs.SETTINGS) {
            return settingsRenderer;
        }
        return songsRenderer;
    }

    private static final class ButtonState {
        final android.widget.Button button;

        ButtonState(android.widget.Button button) {
            this.button = button;
        }
    }

    static final class PreviewState {
        final int scrollY;
        final int generation;
        final SongsRenderer.BatchState batchState;
        final android.widget.Button sourcePlayButton;

        PreviewState(int scrollY, int generation, SongsRenderer.BatchState batchState,
                android.widget.Button sourcePlayButton) {
            this.scrollY = scrollY;
            this.generation = generation;
            this.batchState = batchState;
            this.sourcePlayButton = sourcePlayButton;
        }
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/MainRenderer.kt
package com.dumuzeyn.mp3player

import android.os.Trace
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver
import android.widget.Button
import android.widget.LinearLayout
import android.widget.ScrollView

/** Owns tab rendering, scroll restoration, and the reusable Home hierarchy. */
internal class MainRenderer(private val host: MainActivityCore) {
    private val homeRenderer = HomeMenuRenderer(host)
    private val songsRenderer = SongsMenuRenderer(host)
    private val favoritesRenderer = FavoritesMenuRenderer(host)
    private val playlistsRenderer = PlaylistsMenuRenderer(host)
    private val soundRenderer: MenuRenderer = SoundMenuRenderer(host)
    private val genresRenderer: MenuRenderer = GenresMenuRenderer(host)
    private val artistsRenderer: MenuRenderer = ArtistsMenuRenderer(host)
    private val albumsRenderer: MenuRenderer = AlbumsMenuRenderer(host)
    private val settingsRenderer: MenuRenderer = SettingsMenuRenderer(host)
    private val foldersRenderer: MenuRenderer = FoldersMenuRenderer(host)
    private val scrollPositions = HashMap<String, Int>()
    private val cachedHomeRows = SongRowStateRegistry()
    private var cachedHomeContent: LinearLayout? = null
    private var cachedHomeModel: HomeContent? = null
    private var cachedHomeKey = ""
    private var renderedMenuKey: String? = null

    fun render() {
        rememberCurrentScrollPosition()
        host.refreshTabs()
        host.navigationState.songRenderGeneration++
        if (
            host.navigationState.tabIndex == LibraryTabs.SONGS &&
            !host.navigationState.renderingTabPreview &&
            host.songsView != null
        ) {
            host.songRows.clear()
            host.sourcePlayButton = null
            host.contentScroll?.visibility = View.GONE
            songsRenderer.render()
            renderedMenuKey = menuKey(host.navigationState.tabIndex, host.navigationState.search)
            host.playerUiController.updateMini()
            return
        }
        host.songsView?.hide()
        host.contentScroll?.visibility = View.VISIBLE
        if (host.navigationState.tabIndex == LibraryTabs.HOME && attachCachedHome()) {
            restoreCurrentScrollPosition()
            host.playerUiController.updateMini()
            return
        }
        host.list.removeAllViews()
        host.songRows.clear()
        host.sourcePlayButton = null
        val renderer = rendererForTab()
        val scrollY = scrollPositionFor(host.navigationState.tabIndex, host.navigationState.search)
        if (
            host.navigationState.tabIndex == LibraryTabs.SONGS ||
            host.navigationState.tabIndex == LibraryTabs.FAVORITES
        ) {
            host.songsRenderer.prepareNextRenderForScroll(scrollY)
        }
        if (host.navigationState.tabIndex == LibraryTabs.HOME) {
            renderAndCacheHome(renderer)
        } else {
            host.renderSectionHeader()
            renderer.render()
            if (renderer.needsMiniSpacer()) host.addMiniSpacerIfNeeded()
        }
        restoreCurrentScrollPosition()
        host.playerUiController.updateMini()
    }

    fun captureScrollBeforeUiRebuild() {
        rememberCurrentScrollPosition()
        renderedMenuKey = null
        invalidateHomeCache()
    }

    fun captureCurrentScrollPosition() = rememberCurrentScrollPosition()

    private fun rememberCurrentScrollPosition() {
        val key = renderedMenuKey ?: return
        val scroll = host.contentScroll ?: return
        if (key.startsWith("${LibraryTabs.SONGS}\n")) return
        scrollPositions[key] = scroll.scrollY.coerceAtLeast(0)
    }

    private fun restoreCurrentScrollPosition() {
        val targetKey = menuKey(host.navigationState.tabIndex, host.navigationState.search)
        renderedMenuKey = targetKey
        val targetScroll = host.contentScroll ?: return
        val scrollY = scrollPositions[targetKey]?.coerceAtLeast(0) ?: 0
        if (scrollY <= 0) {
            targetScroll.scrollTo(0, 0)
            targetScroll.visibility = View.VISIBLE
            return
        }
        targetScroll.visibility = View.INVISIBLE
        targetScroll.viewTreeObserver.addOnPreDrawListener(
            object : ViewTreeObserver.OnPreDrawListener {
                override fun onPreDraw(): Boolean {
                    val observer = targetScroll.viewTreeObserver
                    if (observer.isAlive) observer.removeOnPreDrawListener(this)
                    if (
                        host.contentScroll !== targetScroll ||
                        targetKey != menuKey(
                            host.navigationState.tabIndex,
                            host.navigationState.search,
                        )
                    ) {
                        targetScroll.visibility = View.VISIBLE
                        return true
                    }
                    targetScroll.scrollTo(0, scrollY)
                    targetScroll.visibility = View.VISIBLE
                    return false
                }
            },
        )
    }

    private fun menuKey(tabIndex: Int, search: String?): String =
        "$tabIndex\n${search.orEmpty()}"

    fun renderPreview(
        target: LinearLayout,
        targetIndex: Int,
        targetSearch: String?,
    ): PreviewState {
        Trace.beginSection("Voltune/Home.renderTabPreview")
        try {
            val previousList = host.list
            val previousButton = ButtonState(host.sourcePlayButton)
            val previousBatchState = host.songsRenderer.captureBatchState()
            val previousTab = host.navigationState.tabIndex
            val previousGeneration = host.navigationState.songRenderGeneration
            val previousSearch = host.navigationState.search
            val previousPreview = host.navigationState.renderingTabPreview
            val scrollY = scrollPositionFor(targetIndex, targetSearch)
            try {
                host.previewSongRows.clear()
                host.list = target
                host.navigationState.tabIndex = targetIndex
                host.navigationState.songRenderGeneration = previousGeneration + 1
                host.navigationState.search = targetSearch.orEmpty()
                host.navigationState.renderingTabPreview = true
                host.sourcePlayButton = null
                target.removeAllViews()
                val renderer = rendererForTab(targetIndex)
                if (targetIndex == LibraryTabs.SONGS || targetIndex == LibraryTabs.FAVORITES) {
                    host.songsRenderer.prepareNextRenderForScroll(scrollY)
                }
                if (targetIndex == LibraryTabs.HOME && attachCachedHome()) {
                    // Reuse the detached Home hierarchy, artwork, and playback row bindings.
                } else if (targetIndex == LibraryTabs.HOME) {
                    renderAndCacheHome(renderer)
                } else {
                    host.renderSectionHeader()
                    renderer.render()
                    if (renderer.needsMiniSpacer()) host.addMiniSpacerIfNeeded()
                }
                return PreviewState(
                    scrollY,
                    host.navigationState.songRenderGeneration,
                    host.songsRenderer.captureBatchState(),
                    host.sourcePlayButton,
                )
            } finally {
                host.list = previousList
                host.navigationState.tabIndex = previousTab
                host.navigationState.songRenderGeneration = previousGeneration
                host.navigationState.search = previousSearch
                host.navigationState.renderingTabPreview = previousPreview
                host.sourcePlayButton = previousButton.button
                host.songsRenderer.restoreBatchState(previousBatchState)
            }
        } finally {
            Trace.endSection()
        }
    }

    fun adoptPreview(targetIndex: Int, targetSearch: String?, state: PreviewState) {
        Trace.beginSection("Voltune/Home.adoptPreview")
        try {
            host.songsView?.hide()
            renderedMenuKey = menuKey(targetIndex, targetSearch)
            host.navigationState.songRenderGeneration = state.generation
            host.songsRenderer.restoreBatchState(state.batchState)
            host.songRows.replaceWith(host.previewSongRows)
            host.previewSongRows.clear()
            host.sourcePlayButton = state.sourcePlayButton
            if (targetIndex != LibraryTabs.HOME) {
                host.songRows.refresh(host.songRowStateResolver())
            }
            host.artworkUi.promoteVisibleArtwork()
            host.playerUiController.updateMini()
        } finally {
            Trace.endSection()
        }
    }

    fun discardPreview() = host.previewSongRows.clear()

    fun invalidateHomeCache() {
        cachedHomeContent = null
        cachedHomeModel = null
        cachedHomeKey = ""
        cachedHomeRows.clear()
    }

    fun refreshHomePlayback() {
        if (cachedHomeContent?.isAttachedToWindow == true) homeRenderer.refreshPlaybackSection()
    }

    fun setHomeTransitionPaused(paused: Boolean) = homeRenderer.setPlaybackTransitionPaused(paused)

    private fun attachCachedHome(): Boolean {
        val content = cachedHomeContent ?: return false
        if (cachedHomeModel !== host.libraryState.homeContent || cachedHomeKey != homeKey()) {
            return false
        }
        Trace.beginSection("Voltune/Home.cacheHit")
        try {
            val parent = content.parent
            if (parent is ViewGroup && parent !== host.list) parent.removeView(content)
            if (content.parent !== host.list) {
                host.list.removeAllViews()
                host.list.addView(content, LinearLayout.LayoutParams(-1, -2))
            }
            host.activeSongRows().replaceWith(cachedHomeRows)
            host.activeSongRows().refresh(host.songRowStateResolver())
            homeRenderer.refreshPlaybackSection()
            host.sourcePlayButton = null
            return true
        } finally {
            Trace.endSection()
        }
    }

    private fun renderAndCacheHome(renderer: MenuRenderer) {
        Trace.beginSection("Voltune/Home.cacheMiss")
        try {
            val target = host.list
            val content = LinearLayout(host).apply { orientation = LinearLayout.VERTICAL }
            target.addView(content, LinearLayout.LayoutParams(-1, -2))
            host.list = content
            try {
                host.renderSectionHeader()
                renderer.render()
                if (renderer.needsMiniSpacer()) host.addMiniSpacerIfNeeded()
            } finally {
                host.list = target
            }
            cachedHomeContent = content
            cachedHomeModel = host.libraryState.homeContent
            cachedHomeKey = homeKey()
            cachedHomeRows.replaceWith(host.activeSongRows())
        } finally {
            Trace.endSection()
        }
    }

    private fun homeKey(): String = buildString {
        append(host.appearanceState.language)
        append('|').append(host.appearanceState.circularCovers)
        append('|').append(host.panel)
        append('|').append(host.primaryText)
        append('|').append(host.secondaryText)
        append('|').append(host.appearanceState.playlistCardOpacity)
        append('|').append(host.appearanceState.artistCardOpacity)
        append('|').append(host.appearanceState.albumCardOpacity)
    }

    private fun scrollPositionFor(tabIndex: Int, search: String?): Int =
        scrollPositions[menuKey(tabIndex, search)]?.coerceAtLeast(0) ?: 0

    private fun rendererForTab(): MenuRenderer = rendererForTab(host.navigationState.tabIndex)

    private fun rendererForTab(tabIndex: Int): MenuRenderer = when (tabIndex) {
        LibraryTabs.HOME -> homeRenderer
        LibraryTabs.SONGS -> songsRenderer
        LibraryTabs.FAVORITES -> favoritesRenderer
        LibraryTabs.PLAYLISTS -> playlistsRenderer
        LibraryTabs.SOUND -> soundRenderer
        LibraryTabs.GENRES -> genresRenderer
        LibraryTabs.ARTISTS -> artistsRenderer
        LibraryTabs.ALBUMS -> albumsRenderer
        LibraryTabs.FOLDERS -> foldersRenderer
        LibraryTabs.SETTINGS -> settingsRenderer
        else -> songsRenderer
    }

    private class ButtonState(@JvmField val button: Button?)

    class PreviewState(
        @JvmField val scrollY: Int,
        @JvmField val generation: Int,
        @JvmField val batchState: SongsRenderer.BatchState,
        @JvmField val sourcePlayButton: Button?,
    )
}
